Add DSR model

phcarval opened the pull request:
# Description

- Implementation of the DSR model, from https://github.com/VitjanZ/DSR_anomaly_detection 
- References #1121
- Architecture works without crashes (AFAIK), but might not yield the expected performance
- Pretraining needs weights : automatic weight download needs to be set up (unsure how, it's not on TIMM)
- Third phase of training not implemented yet (unsure of how to preceed)
- Still various pre-commit warnings flaring up

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).

phcarval commented:
Added automatic pretrained weight downloading. Now the model can train on its own, but the third phase of training is still missing.
These additions would benefit from being integrated to download.py (as well as the anomaly generator should be added to the already existing files... but this is more complex since DSR has its own way of generating anomalies).

phcarval commented:
I tested the DSR model by loading the authors' pretrained weights and finding the exact same results as them. The following changes are necessary to reproduce :
- `load_state_dict()` for each module in `DsrModel.__init__()`
- normalization: none in config.yaml
- normalization_method: none in config.yaml
- eval batch size is 1
- comment `image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)` in data/utils/image.py
- set `weight_file` to None in `test.py`

Still training left to test as well as implementing the third phase of training.

phcarval commented:
Training seems to work on my end. Only thing left to implement is the upsampling module.

phcarval commented:
DSR is, AFAIK, fully functional right now. Training, testing is implemented for both phases of training. Training procedure from the paper has been copied (just need to adapt the nb of epochs or nb of steps in the config file). There are several small corrections that should be made now to clean up the code.

Let me know if there are problems on your end.

phcarval commented:
I made it so that the scheduler can accept max_steps to anneal the learning rate. BTW, I believe you have an error in EfficientAD lightning_model.py L.188, since the StepLR scheduler's step_size argument sets the number of epochs: instead, you're giving it the number of actual steps. And I guess max() could pose a problem if you give positive max_steps and max_epochs, but one is supposed to be negative?

ashwinvaidya17 reviewed (commented):
Thanks for all the efforts! I have a few comments

samet-akcay commented on src/anomalib/models/dsr/lightning_model.py:
```diff
@@ -0,0 +1,175 @@
+            dict[str, torch.optim.Optimizer | torch.optim.LRScheduler]: Dictionary of optimizers
+        (the first one having a schedule)
+        """
+        num_steps = max(
```
Yeah, when I run `python tools/train.py --model dsr`, I get the following error:

```bash
Traceback (most recent call last):
  File "<string>", line 2, in <module>
TypeError: len() takes exactly one argument (2 given)
```

phcarval replied:
Yes indeed, did not test that commit, thanks.

samet-akcay commented on src/anomalib/models/dsr/lightning_model.py:
```diff
@@ -0,0 +1,175 @@
+            dict[str, torch.optim.Optimizer | torch.optim.LRScheduler]: Dictionary of optimizers
+        (the first one having a schedule)
+        """
+        num_steps = max(
+            self.trainer.max_steps // len(self.trainer.datamodule.train_dataloader(), self.trainer.max_epochs)
+        )
```
```suggestion
        num_steps = max(
            self.trainer.max_steps // len(self.trainer.datamodule.train_dataloader()), self.trainer.max_epochs
        )
```

samet-akcay commented on src/anomalib/models/dsr/loss.py:
```diff
@@ -0,0 +1,81 @@
+
+        self.focal_loss = FocalLoss(alpha=1, reduction="mean")
+
+    def forward(self, gen_mask: Tensor, anomaly_mask: Tensor) -> Tensor:
```
```suggestion
    def forward(self, pred_mask: Tensor, true_mask: Tensor) -> Tensor:
```

phcarval commented on src/anomalib/models/dsr/lightning_model.py:
```diff
@@ -0,0 +1,175 @@
+            dict[str, torch.optim.Optimizer | torch.optim.LRScheduler]: Dictionary of optimizers
+        (the first one having a schedule)
+        """
+        num_steps = max(
+            self.trainer.max_steps // len(self.trainer.datamodule.train_dataloader(), self.trainer.max_epochs)
+        )
+        self.second_phase = int(10 * num_steps / 12)
+        anneal = int(0.8 * self.second_phase)
```
I remember trying to move this to `on_train_start`. I think `configure_optimizers` is called first so this creates an issue?

samet-akcay reviewed (commented):
@phcarval, thanks again for you effort. I've got few comments after my initial round of review, and a question:

I'm training the model now. However, the performance does not seem to be improving. For example, I'm on epoch 272 now, and the performance is the following since the first few epochs:

```
pixel_F1Score=0.109, pixel_AUROC=0.442
```

I see that the total number of epochs is set to 700. Is it improving by then? :)

phcarval commented on src/anomalib/models/dsr/torch_model.py:
```diff
@@ -0,0 +1,1170 @@
+            )
+
+            # reminder : top = lo, bot = hi!
+            return recon_defect, anomaly_mask, quantized_t, quantized_b, anomaly_embedding_lo, anomaly_embedding_hi
```
Done. Can't test it right now though, but I double checked everything.

phcarval commented on src/anomalib/models/dsr/download_weights.py:
```diff
@@ -0,0 +1,119 @@
+"""Pretrained weights downloader for DSR model implementation."""
+
```
I'm going to refactor this, unless you add the weights to Anomalib?

phcarval commented on src/anomalib/models/dsr/config.yaml:
```diff
@@ -0,0 +1,105 @@
+model:
+  name: dsr
+  lr: 0.0002
+  ckpt_path: ./src/anomalib/models/dsr/vq_model_pretrained_128_4096.pckl
```
Sure. I did not know it worked that way, actually. I'm waiting to see if @ashwinvaidya17 adds the weights to Anomalib before changing it.

phcarval commented on src/anomalib/models/dsr/download_weights.py:
Same as above.

phcarval commented:
> @phcarval, thanks again for you effort. I've got few comments after my initial round of review, and a question:
> 
> I'm training the model now. However, the performance does not seem to be improving. For example, I'm on epoch 272 now, and the performance is the following since the first few epochs:
> 
> ```
> pixel_F1Score=0.109, pixel_AUROC=0.442
> ```
> 
> I see that the total number of epochs is set to 700. Is it improving by then? :)

Yes, it's normal since we're not training the upsampling module yet. Your values are actually very similar to mine.
![image](https://github.com/openvinotoolkit/anomalib/assets/31983398/28b1d732-f3e9-4ef4-a020-1892f877ddbe)

However, you should see the image metrics improving.

phcarval commented:
> @phcarval the weights are now part of a release https://github.com/openvinotoolkit/anomalib/releases/tag/dsr_pretrained_weights. Let me know if this works for you.

Great, thanks! However, we only need `checkpoints/vq_model_pretrained_128_4096.pckl` from the zip file. Could you change the release to only include that file? This would also allow me to use the already existing function `extract_and_download()`.

ashwinvaidya17 commented:
> > @phcarval the weights are now part of a release https://github.com/openvinotoolkit/anomalib/releases/tag/dsr_pretrained_weights. Let me know if this works for you.
> 
> Great, thanks! However, we only need `checkpoints/vq_model_pretrained_128_4096.pckl` from the zip file. Could you change the release to only include that file? This would also allow me to use the already existing function `extract_and_download()`.

I have updated the release. Lmk

phcarval commented:
I am now using the existing download function. I have deleted `DsrWeightDownloader`. I have also made minor fixes to EfficientAD because now there are two methods using the `pre_trained/` folder. I have also fixed the epoch length problem for the EfficientAD optimizer, as it is similar to DSR.

phcarval commented:
It turns out that if I set `max_steps` to 120000 for example in the config file, training stops much earlier than expected. On "bottle", training stops at 2222 epochs when it should run to 4444 epochs with a train batch size of 8. Thus, the upsampling module and the scheduler are not attained during training. The training time is basically divided by two. Could this be because of the fact that we use two optimizers, so it counts as two steps even though only one is effectively used at a time?

Note that this problem does not arise when using `max_epochs`.

phcarval commented:
Another issue: the score output by DSR is not computed from the upsampled map, but from the original map. That means that if the upsampling module is performing poorly but the anomaly detection module works well, we will have a very good image-level score but a very poor pixel-level score. When looking at the results folder, it becomes impossible to see which images were actually correctly classified and which images just show poor segmentation performance.

phcarval commented:
Added and tested batch evaluation (`max` -> `amax`). For now, everything OK but the step count and the actual decision not showing in the output images.

samet-akcay commented:
@ashwinvaidya17, @djdameln can we prioritise this please? It's been outstanding for a bit here. Thanks for your patience @phcarval

phcarval commented:
Hi, did you have some time to review the PR? :)
I have merged the main branch. The issues I mentioned still exist but I'm waiting on your input to fix them.

ashwinvaidya17 reviewed (approved):
Sorry for the late review. Thanks again for the efforts!

samet-akcay commented:
Thanks @phcarval, I'll have a look at the issues. Just to double check, is your description up-to-date? I want to know if you have addressed any of the issues you mentioned there.

phcarval commented:
> Thanks @phcarval, I'll have a look at the issues. Just to double check, is your description up-to-date? I want to know if you have addressed any of the issues you mentioned there.

Oh no actually, I didn't update it. The two issues I still are the following :


> It turns out that if I set `max_steps` to 120000 for example in the config file, training stops much earlier than expected. On "bottle", training stops at 2222 epochs when it should run to 4444 epochs with a train batch size of 8. Thus, the upsampling module and the scheduler are not attained during training. The training time is basically divided by two. Could this be because of the fact that we use two optimizers, so it counts as two steps even though only one is effectively used at a time?
> 
> Note that this problem does not arise when using `max_epochs`.



> Another issue: the score output by DSR is not computed from the upsampled map, but from the original map. That means that if the upsampling module is performing poorly but the anomaly detection module works well, we will have a very good image-level score but a very poor pixel-level score. When looking at the results folder, it becomes impossible to see which images were actually correctly classified and which images just show poor segmentation performance.
> 
> I would suggest adding the classification results to each output image. Of course, for most methods it would show redundant information that is technically already present through the output segmentation mask.

phcarval commented:
Readded the path check that was removed in commit d2b6e5a

phcarval commented:
@samet-akcay have you had the time to look at the issues?

samet-akcay commented:
> @samet-akcay have you had the time to look at the issues?

I was out of office, will check this out this week. I'll let you know if I have any update

blaz-r commented:
Hello @phcarval, this all seems like really nice work 😄. 
About your issues: one thing I noticed in the [lightning documentation](https://pytorch-lightning.readthedocs.io/en/1.4.9/common/optimizers.html#automatic-optimization) is that optimizer.step is indeed called for every optimizer (looks like even if None is returned but I'm not sure):
```python
for epoch in epochs:
    for batch in data:
        for opt in optimizers:

            def closure():
                loss = model.training_step(batch, batch_idx, optimizer_idx)
                opt.zero_grad()
                loss.backward()
                return loss

            opt.step(closure)

    for lr_scheduler in lr_schedulers:
        lr_scheduler.step()
```
And based on this https://github.com/Lightning-AI/lightning/issues/16143 I'd say that count goes up for every step call.
That's probably why you see half of max_steps.
[Later in the docs](https://pytorch-lightning.readthedocs.io/en/1.4.9/common/optimizers.html#step-optimizers-at-arbitrary-intervals) they mention that if you want to optimize at custom times, you should override `optimizer_step()` function. I think this might be a proper solution in this case.

Regarding the other thing about the classification and segmentation discrepancy. From the paper it seems that for some cases (transistor) the difference might be quite significant, and I'm just interested if you have any results (metrics and visuals) on that?
From what you say, and also for other cases, I do think that the visualization should also include the anomaly score (which I think actually is somewhat included in some inferencers like OpenVINO). This would probably need a separate PR.

phcarval commented:
Hi @blaz-r ! Thanks a lot for the comments.

The `optimizer_step()` function indeed looks interesting. I guess it does advance one step for each optimizer, even if returning None, so I'll look into using that.

I have no specific results to show: rather, this problem stems from the optimizer issue. If, for some reason, someone does not train the upsampling module for enough steps, they should be able to see which images were correctly classified. In that case, I agree, another PR should be issued for that.

phcarval commented:
I have the same problem as https://github.com/Lightning-AI/lightning/discussions/10893: I cannot make an "empty" optimizer step, so the step count increases for each optimizer.

phcarval commented:
I fixed it using manual optimization. However, it doesn't improve maintainability and perhaps it should be changed if we find a way to make it work using core Lightning functionality.

blaz-r commented:
Nice 😄. That's actually a good solution, in my opinion even a more clean one. Good job.
We discussed the visualization, and the idea is to add a label with the image name and anomaly score to the current visualizer. That would of course be a separate PR. Does that sound like something you had in mind?

phcarval commented:
> Nice 😄. That's actually a good solution, in my opinion even a more clean one. Good job. 

Thanks for the feedback!

> We discussed the visualization, and the idea is to add a label with the image name and anomaly score to the current visualizer. That would of course be a separate PR. Does that sound like something you had in mind?

Indeed, this proposition sounds fine to me.

samet-akcay commented:
> I fixed it using manual optimization. However, it doesn't improve maintainability and perhaps it should be changed if we find a way to make it work using core Lightning functionality.

Thanks for the fix @phcarval! I agree with @blaz-r. Your solution also shows that this could be done manually via vanilla torch, which demonstrates flexibility. I think it would be a good example to some users/developers for future implementations.

phcarval commented:
> > I fixed it using manual optimization. However, it doesn't improve maintainability and perhaps it should be changed if we find a way to make it work using core Lightning functionality.
> 
> Thanks for the fix @phcarval! I agree with @blaz-r. Your solution also shows that this could be done manually via vanilla torch, which demonstrates flexibility. I think it would be a good example to some users/developers for future implementations.

Thanks for the feedback! My concern was that, AFAIK, it is the only method that uses a different optimization system in Anomalib and so it could create somewhat of an issue down the line if the method is refactored, as optimization is supposed to be "automatic" in Lightning. It's probably a very minor problem though.

Anyway, I have no more outstanding issues on my end!

samet-akcay commented on src/anomalib/models/dsr/torch_model.py:
```diff
@@ -0,0 +1,1196 @@
+    Args:
+        embedding_dim (int): Dimension of codebook embeddings.
+        num_embeddings (int): Number of embeddings.
+        anom_par (float):
```
Is this intentionally left blank?

samet-akcay replied:
Also, as raised above, `anom_par` is not so clear to understand. It would be great if it is renamed

phcarval replied:
See above

samet-akcay commented on src/anomalib/models/dsr/torch_model.py:
```diff
@@ -0,0 +1,1196 @@
+                - Computed segmentation mask (M)
+                - Resized ground-truth anomaly map (M_gt)
+        """
+        # top == lo
```
left over?

samet-akcay commented on src/anomalib/models/dsr/torch_model.py:
```diff
@@ -0,0 +1,1196 @@
+        """
+        # top == lo
+
+        outputs = None
```
It would be more readable if `outputs` is annotated. Something like 
```suggestion
        outputs: dict[str, Tensor]
```

samet-akcay commented on src/anomalib/models/dsr/torch_model.py:
```diff
@@ -0,0 +1,1196 @@
+                "embedding_top": embd_top,
+                "obj_spec_image": spec_image_def,
+                "pred_mask": out_mask_sm,
+                "true_mask": anomaly_map,
```
is `anomaly_map` really a `true_mask`?

samet-akcay reviewed (changes requested):
Thanks @phcarval, this is a huge PR :)

I've got some minor comments, but the overall code quality is quite high! Thanks again for your contribution and patience!

samet-akcay commented:
By the way, there is one failed test, mainly because `DsrModel` returns a `dict`. We probably need a separate PR to handle torch outputs that are dict

phcarval commented on docs/source/reference_guide/algorithms/dsr.rst:
```diff
@@ -0,0 +1,45 @@
+Architecture
+************
+
+.. image:: ../../images/dsr/architecture.png
```
The link doesn't work as of now, is this intended?

samet-akcay replied:
it will work once we merge this

phcarval commented on src/anomalib/models/dsr/README.md:
```diff
@@ -0,0 +1,21 @@
+
+## Architecture
+
+![DSR Architecture](../../../docs/source/images/dsr/architecture.png "DSR Architecture")
```
As above

phcarval commented on src/anomalib/models/dsr/config.yaml:
```diff
@@ -0,0 +1,105 @@
+model:
+  name: dsr
+  lr: 0.0002
+  anom_par: 0.2
```
I suggest `anomaly_strength` or `latent_anomaly_strength`

phcarval commented on src/anomalib/models/dsr/lightning_model.py:
```diff
@@ -0,0 +1,184 @@
+        self.quantized_anomaly_generator = DsrAnomalyGenerator()
+        self.perlin_generator = Augmenter()
+        self.model = DsrModel(anom_par)
+        self.second_loss = DsrSecondLoss()
+        self.third_loss = DsrThirdLoss()
```
Will do

phcarval commented:
I went though all requested changes @samet-akcay !

samet-akcay commented on src/anomalib/models/dsr/lightning_model.py:
```diff
@@ -0,0 +1,184 @@
+        del args, kwargs  # These variables are not used.
+
+        model_outputs = self.model(batch["image"])
+        batch["anomaly_maps"] = model_outputs["pred_mask"]
+        batch["pred_scores"] = model_outputs["score"]
```
Is it possible if `model_outputs` have the same keys for `anomaly_maps` and `pred_scores`? The reason for asking is that we need standard keys for the `TorchInferencer` stuff?

The tests are failing since the output model is returning a different output than what is expected here
https://github.com/openvinotoolkit/anomalib/blob/5eb20c64cc6ee8862bf2723b83fc4cae60a5594e/src/anomalib/deploy/inferencers/torch_inferencer.py#L155-L185

samet-akcay replied:
I have create PR #1354 to handle cases like this. `DsrModel` needs to return `anomaly_map` and `pred_score` though.

phcarval commented:
@samet-akcay have you had the time to test the model? I seem to have very bad segmentation results, this might be due to the optimizer change. I would like to know if you had that problem too.

phcarval commented:
When removing the default imagenet normalization parameter, training seems to be going much better and my results are close to the paper's. Only problem is the one I describe in issue https://github.com/openvinotoolkit/anomalib/issues/1380 where images are segmented even if they're classified as non-anomalous.

samet-akcay reviewed (approved):
Thanks @phcarval, looking great! Really appreciate your patience, and thanks again for this great contribution!
